File: OpenManus/langgraph_agent/tools/browser_tool.py
from typing import Optional, Callable, Awaitable
from langchain_core.tools import tool
from pydantic import BaseModel, Field
import asyncio


class BrowserAction(BaseModel):
    action: str = Field(description="Action: navigate, click, type, scroll, screenshot, extract_text")
    url: Optional[str] = Field(default=None, description="URL for navigate action")
    selector: Optional[str] = Field(default=None, description="CSS selector for click/type actions")
    text: Optional[str] = Field(default=None, description="Text for type action")
    direction: Optional[str] = Field(default=None, description="Scroll direction: up, down")


class BrowserTools:
    def __init__(self, on_screenshot: Optional[Callable[[str], Awaitable[None]]] = None):
        self._browser = None
        self._context = None
        self._page = None
        self.on_screenshot = on_screenshot
    
    async def _ensure_browser(self):
        if self._browser is None:
            from playwright.async_api import async_playwright
            self._playwright = await async_playwright().start()
            self._browser = await self._playwright.chromium.launch(headless=True)
            self._context = await self._browser.new_context(
                viewport={"width": 1280, "height": 720}
            )
            self._page = await self._context.new_page()
        return self._page

    async def navigate(self, url: str) -> dict:
        page = await self._ensure_browser()
        await page.goto(url, wait_until="domcontentloaded", timeout=30000)
        screenshot = await page.screenshot(type="png")
        import base64
        screenshot_b64 = base64.b64encode(screenshot).decode()
        if self.on_screenshot:
            await self.on_screenshot(screenshot_b64)
        return {
            "success": True,
            "url": page.url,
            "title": await page.title(),
            "screenshot": screenshot_b64
        }

    async def click(self, selector: str) -> dict:
        page = await self._ensure_browser()
        try:
            await page.click(selector, timeout=5000)
            await page.wait_for_load_state("domcontentloaded", timeout=10000)
            screenshot = await page.screenshot(type="png")
            import base64
            screenshot_b64 = base64.b64encode(screenshot).decode()
            if self.on_screenshot:
                await self.on_screenshot(screenshot_b64)
            return {"success": True, "screenshot": screenshot_b64}
        except Exception as e:
            return {"success": False, "error": str(e)}

    async def type_text(self, selector: str, text: str) -> dict:
        page = await self._ensure_browser()
        try:
            await page.fill(selector, text, timeout=5000)
            screenshot = await page.screenshot(type="png")
            import base64
            screenshot_b64 = base64.b64encode(screenshot).decode()
            if self.on_screenshot:
                await self.on_screenshot(screenshot_b64)
            return {"success": True, "screenshot": screenshot_b64}
        except Exception as e:
            return {"success": False, "error": str(e)}

    async def screenshot(self) -> dict:
        page = await self._ensure_browser()
        screenshot = await page.screenshot(type="png")
        import base64
        screenshot_b64 = base64.b64encode(screenshot).decode()
        if self.on_screenshot:
            await self.on_screenshot(screenshot_b64)
        return {
            "success": True,
            "url": page.url,
            "title": await page.title(),
            "screenshot": screenshot_b64
        }

    async def extract_text(self) -> dict:
        page = await self._ensure_browser()
        text = await page.evaluate("document.body.innerText")
        return {"success": True, "text": text[:5000]}

    async def scroll(self, direction: str = "down") -> dict:
        page = await self._ensure_browser()
        delta = 500 if direction == "down" else -500
        await page.mouse.wheel(0, delta)
        await asyncio.sleep(0.5)
        screenshot = await page.screenshot(type="png")
        import base64
        screenshot_b64 = base64.b64encode(screenshot).decode()
        if self.on_screenshot:
            await self.on_screenshot(screenshot_b64)
        return {"success": True, "screenshot": screenshot_b64}

    async def execute(self, action: BrowserAction) -> dict:
        if action.action == "navigate" and action.url:
            return await self.navigate(action.url)
        elif action.action == "click" and action.selector:
            return await self.click(action.selector)
        elif action.action == "type" and action.selector and action.text:
            return await self.type_text(action.selector, action.text)
        elif action.action == "screenshot":
            return await self.screenshot()
        elif action.action == "extract_text":
            return await self.extract_text()
        elif action.action == "scroll":
            return await self.scroll(action.direction or "down")
        else:
            return {"success": False, "error": f"Unknown action: {action.action}"}

    async def cleanup(self):
        if self._browser:
            await self._browser.close()
        if hasattr(self, '_playwright') and self._playwright:
            await self._playwright.stop()


def create_browser_tool(browser_tools: BrowserTools):
    @tool
    async def browser_use(
        action: str,
        url: Optional[str] = None,
        selector: Optional[str] = None, 
        text: Optional[str] = None,
        direction: Optional[str] = None
    ) -> str:
        """Control a web browser to navigate, click, type, scroll, or extract content.
        
        Args:
            action: One of: navigate, click, type, scroll, screenshot, extract_text
            url: URL to navigate to (for navigate action)
            selector: CSS selector (for click/type actions)
            text: Text to type (for type action)
            direction: Scroll direction: up or down (for scroll action)
        """
        browser_action = BrowserAction(
            action=action,
            url=url,
            selector=selector,
            text=text,
            direction=direction
        )
        result = await browser_tools.execute(browser_action)
        if "screenshot" in result:
            result_copy = result.copy()
            result_copy["screenshot"] = "[screenshot captured]"
            return str(result_copy)
        return str(result)
    
    return browser_use

File: alfie/alfie-backend/src/agent/tools/BrowserTool.js
import { BaseTool } from '../BaseTool.js';
import { chromium } from 'playwright';

let browserInstance = null;

async function getBrowser() {
  if (!browserInstance || !browserInstance.isConnected()) {
    browserInstance = await chromium.launch({ headless: true, args: ['--no-sandbox', '--disable-setuid-sandbox'] });
  }
  return browserInstance;
}

export class BrowserTool extends BaseTool {
  constructor() {
    super({
      name: 'browser',
      description: 'Browse websites to extract content, take screenshots, or interact with web pages. Operations: navigate (go to URL and get page text), screenshot (capture page as image), click (click an element), type (type text into an input), extract (extract structured data using CSS selectors).',
      parameters: {
        properties: {
          operation: { type: 'string', enum: ['navigate', 'screenshot', 'click', 'type', 'extract'], description: 'Browser operation' },
          url: { type: 'string', description: 'URL to navigate to (for navigate/screenshot)' },
          selector: { type: 'string', description: 'CSS selector for click/type/extract operations' },
          text: { type: 'string', description: 'Text to type (for type operation)' },
          extractFields: { type: 'object', description: 'Map of field names to CSS selectors (for extract operation)' },
        },
        required: ['operation'],
      },
    });
    this.timeout = 30000;
    this.page = null;
  }

  async execute(input, context) {
    const browser = await getBrowser();
    if (!this.page || this.page.isClosed()) {
      this.page = await browser.newPage();
    }
    
    switch (input.operation) {
      case 'navigate': {
        if (!input.url) throw new Error('url required for navigate');
        await this.page.goto(input.url, { waitUntil: 'domcontentloaded', timeout: 20000 });
        const title = await this.page.title();
        const text = await this.page.evaluate(() => {
          const body = document.body;
          const clone = body.cloneNode(true);
          clone.querySelectorAll('script, style, nav, footer, header').forEach(el => {
            el.remove();
          });
          return clone.innerText?.slice(0, 15000) || '';
        });
        return { title, url: this.page.url(), text };
      }
      case 'screenshot': {
        if (input.url) {
          await this.page.goto(input.url, { waitUntil: 'domcontentloaded', timeout: 20000 });
        }
        const buffer = await this.page.screenshot({ type: 'png', fullPage: false });
        return { screenshot: `data:image/png;base64,${buffer.toString('base64')}`, url: this.page.url() };
      }
      case 'click': {
        if (!input.selector) throw new Error('selector required for click');
        await this.page.click(input.selector, { timeout: 5000 });
        await this.page.waitForTimeout(1000);
        return { clicked: input.selector, url: this.page.url() };
      }
      case 'type': {
        if (!input.selector || !input.text) throw new Error('selector and text required for type');
        await this.page.fill(input.selector, input.text);
        return { typed: input.text, selector: input.selector };
      }
      case 'extract': {
        if (!input.selector && !input.extractFields) throw new Error('selector or extractFields required');
        if (input.extractFields) {
          const results = {};
          for (const [key, sel] of Object.entries(input.extractFields)) {
            results[key] = await this.page.$$eval(sel, els => els.map(el => el.textContent?.trim()).filter(Boolean));
          }
          return results;
        }
        const elements = await this.page.$$eval(input.selector, els => els.map(el => el.textContent?.trim()).filter(Boolean));
        return { elements: elements.slice(0, 100) };
      }
      default:
        throw new Error(`Unknown browser operation: ${input.operation}`);
    }
  }
}
